Game/Source/GDKTestGyms/BenchmarkGymGameModeBase.cpp

```diff
@@ -0,0 +1,371 @@
+	{
+		return;
+	}
+
+	USpatialNetDriver* SpatialDriver = Cast<USpatialNetDriver>(GetNetDriver());
+	if (ensure(SpatialDriver != nullptr))
+	{
+		USpatialMetrics* SpatialMetrics = SpatialDriver->SpatialMetrics;
+		if (ensure(SpatialMetrics != nullptr))
+		{
+			{
+				// Valid on all workers
+				UserSuppliedMetric Delegate;
+				Delegate.BindUObject(this, &ABenchmarkGymGameModeBase::GetFPSValid);
+				SpatialMetrics->SetCustomMetric(AverageFPSValid, Delegate);
+			}
+
+			if (HasAuthority())
+			{
+				{
+					UserSuppliedMetric Delegate;
+					Delegate.BindUObject(this, &ABenchmarkGymGameModeBase::GetClientRTT);
+					SpatialMetrics->SetCustomMetric(AverageClientRTTMetricName, Delegate);
+				}
+
+				{
+					UserSuppliedMetric Delegate;
+					Delegate.BindUObject(this, &ABenchmarkGymGameModeBase::GetClientViewLateness);
+					SpatialMetrics->SetCustomMetric(AverageClientViewLatenessMetricName, Delegate);
+				}
+
+				{
+					UserSuppliedMetric Delegate;
+					Delegate.BindUObject(this, &ABenchmarkGymGameModeBase::GetPlayersConnected);
+					SpatialMetrics->SetCustomMetric(PlayersSpawnedMetricName, Delegate);
+				}
+
+				{
+					UserSuppliedMetric Delegate;
+					Delegate.BindUObject(this, &ABenchmarkGymGameModeBase::GetClientFPSValid);
+					SpatialMetrics->SetCustomMetric(AverageClientFPSValid, Delegate);
+				}
+			}
+		}
+	}
+}
+
+void ABenchmarkGymGameModeBase::Tick(float DeltaSeconds)
+{
+	Super::Tick(DeltaSeconds);
+
+	TickFPSCheck(DeltaSeconds);
+	TickPlayersConnectedCheck(DeltaSeconds);
+	TickUXMetricCheck(DeltaSeconds);
+}
+
+void ABenchmarkGymGameModeBase::TickPlayersConnectedCheck(float DeltaSeconds)
+{
+	if (!HasAuthority())
+	{
+		return;
+	}
+
+	if (SecondsTillPlayerCheck > 0.0f)
+	{
+		SecondsTillPlayerCheck -= DeltaSeconds;
+		if (SecondsTillPlayerCheck <= 0.0f)
+		{
+			if (ActivePlayers != ExpectedPlayers)
+			{
+				// This log is used by the NFR pipeline to indicate if a client failed to connect
+				UE_LOG(LogBenchmarkGymGameModeBase, Error, TEXT("A client connection was dropped. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
+			}
+			else
+			{
+				// Useful for NFR log inspection
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("All clients successfully connected. Expected %d, got %d"), ExpectedPlayers, GetNumPlayers());
+			}
+		}
+	}
+}
+
+void ABenchmarkGymGameModeBase::TickFPSCheck(float DeltaSeconds)
+{
+	const UNFRConstants* Constants = UNFRConstants::Get(GetWorld());
+	check(Constants);
+	if (Constants->SamplesForFPSValid() && !bHasFpsFailed && GetWorld() != nullptr)
+	{
+		if (const UGDKTestGymsGameInstance* GameInstance = Cast<UGDKTestGymsGameInstance>(GetWorld()->GetGameInstance()))
+		{
+			float FPS = GameInstance->GetAveragedFPS();
+			if (FPS < Constants->GetMinServerFPS())
+			{
+				bHasFpsFailed = true;
+#if !WITH_EDITOR
+				UE_LOG(LogBenchmarkGymGameModeBase, Log, TEXT("FPS check failed. FPS: %.8f"), FPS);
+#endif
+			}
+		}
+	}
+
+	bool bClientFpsWasValid = true;
```

Comment:
From here on this only needs to be done on the auth server. Perhaps split this into two functions, one for server (auth and non-auth), and one for client checks (auth only)?